src/mlpack/methods/isomap/k_nearest.hpp

```diff
@@ -0,0 +1,115 @@
+/**
+ * @file k_nearest.hpp
+ * @author Rishabh Ranjan
+ * 
+ * This file defines the class K_Nearest which constructs the neighbourhood
+ * graph for Isomap, according to the number of neighbours provided.
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ISOMAP_K_NEAREST
+#define MLPACK_METHODS_ISOMAP_K_NEAREST
+
+#include <mlpack/prereqs.hpp>
+#include <queue>
+
+namespace mlpack {
+namespace isomap {
+/**
+ * This class implements the K-Nearest Neighbours function to construct
+ * a neighbourhood graph for the given distance matrix.
+*/
+
+class K_Nearest
+{
+ public:
+  /**
+   * Function to make the neighbourhood graph for the given distance matrix
+   * and number of neighbours.
+   * 
+   * @param disMat -distance matrix to construct the graph
+   * @param n_neighbours -number of neighbours to consider
+   */
+  void MakeNeighbourhoodGraph(arma::mat &disMat,
+                              const int n_neighbours)
+  {
+    // go through every row
+    for (size_t i = 0; i < disMat.n_rows; i++)
+    {
+      arma::uvec temp;
+
+      // find indices if row is sorted in descending order
+      temp = sort_index(disMat.row(i), "descend");
+
+      // set distance to inf for every neighbour farther that n_neighbours
+      for (size_t j = 0; j < temp.size()-n_neighbours-1; j++)
+        disMat(i, temp(j)) = DBL_MAX;
+    }
+
+    // check if neighbourhood graph is connected
+    if (!IsConnected(disMat))
+      Log::Fatal << "Constructed neighbourhood graph is not connected. "
+                  << "Increase the number of neighbours (Default is 3).\n";
+  }
```

Comment:
Actually this method can be greatly simplified or removed entirely if we use mlpack's nearest neighbor search infrastructure, since the `Search()` method will return only the k nearest neighbors instead of everything, and they will be in sorted order.